### PR TITLE
Change RawGestureDetector API to be better for strong mode Dart.

### PR DESCRIPTION
**Pull request by Hixie**

### CODE REVIEW

**Comment by Hixie** [comment]
Fixes https://github.com/flutter/flutter/issues/7173

**Comment by HansMuller** [comment]
This seems like a good change, the code that uses raw gesture detectors is easier to read for it.

LGTM although Travis has found a test failure.

File: packages/flutter/lib/src/widgets/gesture_detector.dart [hunk 1]
```diff
@@ -419,9 +531,12 @@ class RawGestureDetectorState extends State<RawGestureDetector> {
     final Map<Type, GestureRecognizer> oldRecognizers = _recognizers;
     _recognizers = <Type, GestureRecognizer>{};
     for (Type type in gestures.keys) {
+      assert(gestures[type] != null);
+      assert(gestures[type]._debugAssertTypeMatches(type));
       assert(!_recognizers.containsKey(type));
-      _recognizers[type] = gestures[type](oldRecognizers[type]);
-      assert(_recognizers[type].runtimeType == type);
+      _recognizers[type] ??= gestures[type].constructor();
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/gesture_detector.dart]: wow this line was completely buggy

**Comment by Hixie** [comment]
Fixed. Will check in on green.

**Pull request merged**